client/service/src/lib.rs

```diff
@@ -608,7 +608,7 @@ where
 	H: std::hash::Hash + Eq + sp_runtime::traits::Member + sp_runtime::traits::MaybeSerialize,
 	E: 'static + IntoPoolError + From<sp_transaction_pool::error::Error>,
 {
-	fn transactions(&self) -> Vec<(H, <B as BlockT>::Extrinsic)> {
+	fn transactions(&self) -> Vec<(H, B::Extrinsic)> {
```

Comment:
Would be really awesome to return `Arc<_>` here as well and avoid cloning. I believe that a `&B::Extrinsic` (acquired through `.data()` method) should be more than enough to construct package encoding. If we reach the limit you set in some other place we would be cloning 50MB of data here, which sounds insane.

Also as discussed in DMs, I think it would be good to introduce some sane upper bound here, like to propagate at most X transactions from the front, `X` possibly being something like `max(poolsize / 4, 1024)`, to be a little bit defensive and avoid potential issues in the future.

Reply:
I will create issue for this, I would like not to change that much here.

Reply:
Also I think limit should depend on last few blocks avg size (in transaction count), but it's quite involving calculating it here, and out of scope of this pr.

Reply:
Same as above, getting an `Arc` does not make much sense.

We could possibly get pending transaction hashes though and pass raw hashes to the networking and let it batch-request transactions from the queue to construct the packet (i.e. clone there).